[RFC] tools/cgexec: add -r option for systemd scope

kamalesh-babulal opened the pull request:
This patch set provides an option for users to replace the default idle_thread
created for systemd scope. It introduces '-r' option, when specified, the program
passed as the argument to cgexec will replace the default idle_thread, if it
exists. Also, if it's the only task, killing it will remove the .scope cgroup it's running
in. It also adds support for ftests to pass the -r flag and a test case to exercise the
different likely and unlikely paths with replace flag.

kamalesh-babulal commented:
Technically this V3 of #341

**V3 changes:**
- Rebased to the latest main
- Fixed the logic of checking both /sys/fs/cgroup/{systemd,[unified]}/ availability
- Minor non-functional improvements

drakenclimber commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+	}
+
+	if (is_cgroup_mode_legacy() && (found_systemd_cgrp == 0 || found_unified_cgrp == 1)) {
+			err("cgroup legacy setup incorrect\n");
```
nitpick - indentation looks weird

kamalesh-babulal replied:
Agreed, will fix all of them.

drakenclimber commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+	}
+
+	if (is_cgroup_mode_hybrid() && (found_systemd_cgrp == 0 || found_unified_cgrp == 0)) {
+			err("cgroup hybrid setup incorrect\n");
```
nitpick - indentation

drakenclimber commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+	int len;
+
+	pthread_rwlock_rdlock(&cg_mount_table_lock);
+	cg_build_path_locked("", systemd_cgroup_name, systemd_ctrl_name);
```
Should the first parameter be `NULL` instead of `""`?

kamalesh-babulal replied:
Sure, will fix it.

drakenclimber commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+	cg_build_path_locked("", systemd_cgroup_name, systemd_ctrl_name);
+	pthread_rwlock_unlock(&cg_mount_table_lock);
+
+	ctrl_name = strstr(systemd_cgroup_name, systemd_ctrl_name);
```
It might be good to add a couple comments of what you expect the string to look like each step of the way here.  I won't remember when I re-look at this in a few months (or days) :)

kamalesh-babulal replied:
Agreed, will add more comments.

drakenclimber commented on tests/ftests/085-sudo-cgexec_systemd_scope.py:
```diff
@@ -0,0 +1,160 @@
+    Cgroup.delete(config, CONTROLLER, OUT_OF_SCOPE_CGNAME1)
+    Cgroup.delete(config, CONTROLLER, OUT_OF_SCOPE_CGNAME2)
+
+    pid = Cgroup.get_pids_in_cgroup(config, CGNAME, CONTROLLER)[0]
```
It's unlikely there are more than one processes in the cgroup, but it might be better to just get the entire list and pass it to Process.kill()

```
pids = Cgroup.get_pids_in_cgroup(config, CGNAME, CONTROLLER)
Process.kill(config, pids)
```

drakenclimber commented on tests/ftests/085-sudo-cgexec_systemd_scope.py:
```diff
@@ -0,0 +1,160 @@
+    Process.kill(config, pid)
+
+    # kill the process, that was created by us but failed to migrate to the cgroup
+    active = active_children()
```
nifty.  I didn't know about this command :)

drakenclimber commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+	/* Let's parse the cgroup of the pid, to check if its in one or
+	 * more .scopes.
+	 */
+	snprintf(buffer, FILENAME_MAX, "/proc/%u/cgroup", pid);
```
No pressure either way, as I realize that could be a large undertaking

kamalesh-babulal commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+	return scope_pid;
+
+}
+static int find_scope_pid(pid_t pid)
```
@drakenclimber Thank you so much for reviewing the patch set. Sure, will fix it.

kamalesh-babulal commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
 
+	while (fgets(buffer, FILENAME_MAX, pid_proc_fp)) {
+		/* read according to the cgroup mode */
+		if (strstr(buffer, "::"))
```
The `sscanf()` for both modes will fail.

kamalesh-babulal commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+	 * not accounting for named hierarchies, which can be more
+	 * than 16 themselves.
+	 */
+	static char parsed_scope_path[16][FILENAME_MAX];
```
Sure, make sense to use `MAX_MNT_ELEMENTS`, and will add a comment.

kamalesh-babulal commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+	return -1;
+}
+
+static int search_systemd_idle_thread_task(pid_t pids[], size_t size)
```
Sure, will do

kamalesh-babulal commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+			continue;
+
+		/* cgroup v1 might have shared mount points cpu,cpuacct */
+		_ctrl_name = strchr(ctrl_name, ',');
```
`ctrl_name` is assigned before reading, I guess it is not possible to read it without an assignment.

kamalesh-babulal commented on src/tools/cgexec.c:
```diff
@@ -171,12 +188,284 @@ int main(int argc, char *argv[])
+		if (strstr(buffer, "::"))
+			ret = sscanf(buffer, "%d::%4096s\n", &idx, cgroup_name);
+		else
+			ret = sscanf(buffer, "%d:%[^:]:%4096s\n", &idx, ctrl_name, cgroup_name);
```
Sure, will fix it.

kamalesh-babulal commented:
**V4 Changes**
- Rebased on the latest main tree
- Addressed comments from #355 from @drakenclimber 
  - fixed memory leak from `write_systemd_unified()`

drakenclimber commented:
Pushed to main.  Thanks, @kamalesh-babulal
